Add DomainRegistry smart contract

getWatermelon opened the pull request:
add DomainRegistry smart contract, add deploy and test script for a DomainRegistry

getWatermelon committed:
add DomainRegistry smart contract, add deploy and test script for a DomainRegistry

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+    uint256 public registrationFee;
+
+    /// @dev Struct to store information about a domain
+    struct Domain {
```
line 15 
this Struct is a little bit over -logic for this task 
you do not need to store timestamp, iit is already stored in the block
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    /// @dev Array of all domains registered in the contract
+    Domain[] public domains;
```
line 22 - 28 
It is expansive to store same data in different places
maybe you can find a way to store the data in one place 
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+        _;
+    }
+
+    /// @notice Creates a new DomainRegistry contract
```
line 43-44 and other same places
NatSpec comments for multiple lines should be written with /** ... */
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+    /// @notice Retrieves the domains controlled by a specific address
+    /// @param _controllerAddress The address of the controller to query domains for
+    /// @return An array of `Domain` structs controlled by `_controllerAddress`
+    function getControllerDomains(address _controllerAddress) public view returns (Domain[] memory) {
```
line 53 
function may be external and save some gas
when you return array as value with small length it will work, but if the array is big it will not work
function should be moved according to NatSpec rules
// - 1 point

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+    /// @notice Changes the registration fee
+    /// @dev Can only be called by the contract owner
+    /// @param _newFee The new fee for registering a domain
+    function changeRegistrationFee(uint256 _newFee) external onlyOwner {
```
line 80
no income data validation
no event for this case
// - 1 point

IhorSolDevRD replied:
uint256 from 0 to max

IhorSolDevRD commented on contracts/DomainRegistry.sol:
```diff
@@ -0,0 +1,89 @@
+
+    /// @notice Withdraws the collected registration fees to the owner's address
+    /// @dev Can only be called by the contract owner
+    function withdrawFees() external onlyOwner {
```
line 86 
it is good practice to use call instead of transfer
// - 1 point

IhorSolDevRD commented on test/DomainRegistry.js:
```diff
@@ -0,0 +1,227 @@
+
+    describe("Deployment", function () {
+        it("Should set the right owner and registration fee", async function () {
+            const { domainRegistry, registrationFee, owner } = await loadFixture(deployDomainRegistryFixture);
```
line 22 and same places
if this line is moved to Before section, it will be more clear
// - 1 point

getWatermelon committed:
resolve issues in domain-registry:
  remove Domain struct
  optimize memory usage
  make getControllerDomains function external
  add pagination functionality to getControllerDomains function with max limit validation
  use before instead of loadFixture in tests
  add multiline NatSpec comments
  add custom errors
  add RegistrationFeeChanged event
  remove timestamp from events
  add validation on registerDomain and changeRegistrationFee functions

getWatermelon committed:
reformat file

getWatermelon committed:
remove unnecessary comments from tests

getWatermelon committed:
remove domain max length limitation

getWatermelon committed:
remove max controller domains search limitation

getWatermelon committed:
add getControllerDomainsNumber function and tests for it

getWatermelon committed:
rework DomainRegistry,
  add domainToHolder mapping instead of controllerToDomains and registeredDomains,
  add registerDomain function
  remove getControllerDomains and getControllerDomainsNumber functions because this data are collected by events
  update tests
